### PR TITLE
File Format for DDLS (CDS Data Definitions)

### PR DESCRIPTION
**Pull request by BeckerWdf**

### CODE REVIEW

File: file-formats/ddls/ddls.json [hunk 1]
```diff
@@ -0,0 +1,126 @@
+
+{ 
+  "$schema": "https://json-schema.org/draft/2020-12/schema", 
+  "$id": "http://sap.com/schema/nrob.json",
```
Comment (Christopher-Hermann) [inline file-formats/ddls/ddls.json:4]: Id should be the actual location of the schema
```suggestion
  "$id": "https://github.com/SAP/abap-file-formats/blob/main/file-formats/ddls/ddls.json",
```

File: file-formats/ddls/ddls.json [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      "additionalProperties": false, 
+      "required": [ 
+        "description", 
+        "originalLanguage" 
+      ] 
+    }, 
+    "sourceOrigin": { 
+      "title": "Source Origin", 
+      "description": "Source Origin", 
+      "type": "integer",
```
Comment (Christopher-Hermann) [inline file-formats/ddls/ddls.json:57]: Only type string is possible for enums. I assume that the schema is generated. If so, we have to adapt the generator.

File: file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (Christopher-Hermann) [inline file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json:1]: Formatting the json would increase the readability

**Review by Christopher-Hermann — CHANGES_REQUESTED** [review]
Probably one point has to be addressed by us.

**Comment by larshp** [comment]
A: as a generic issue, we need to consider what the filename extensions should be, I dont think ".cds" is currently used anywhere?
B: I have mixed thoughts about the BASEINFO
C: generic issue #115, baseinfo is not a LIMU subobject? If its json, then it can go into the main json file, instead of having multiple files?

File: file-formats/ddls/ddls.json (hunk 2 above)
Comment (wurzka) [reply file-formats/ddls/ddls.json]: If you link the type to ty_source_origin to the corresponding constants like it's done for example in [zif_aff_intf_v1](https://github.com/SAP/abap-file-formats/blob/main/file-formats/intf/type/zif_aff_intf_v1.intf.abap), the generator adds the right type

File: file-formats/ddls/ddls.json (hunk 1 above)
Comment (BeckerWdf) [reply file-formats/ddls/ddls.json]: Thanks. Fixed with next commit.

File: file-formats/ddls/ddls.json (hunk 2 above)
Comment (BeckerWdf) [reply file-formats/ddls/ddls.json]: Thanks. Fixed with next commit.

File: file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json (hunk 3 above)
Comment (BeckerWdf) [reply file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json:1]: Let's handle this in a separate PR.

**Comment by schneidermic0** [comment]
>B: I have mixed thoughts about the BASEINFO
>We don't need this for the review use-case but will need it for the "import" use-case.

The `.objectdependencies.json` file (BASEINFO) might be related to issue #131. Does anybody know, whether abapGit, as of today, uses the information stored in the `.objectdependencies.json` file?

File: file-formats/ddls/ddls.json (hunk 2 above)
Comment (BeckerWdf) [reply file-formats/ddls/ddls.json]: > ty_source_origin

Thanks. Will fix.

**Comment by larshp** [comment]
abapGit, yes, it uses baseinfo today: https://github.com/abapGit/abapGit/blob/main/src/objects/zcl_abapgit_object_ddls.clas.abap#L236

**Comment by larshp** [comment]
Currently,

- ADT uses `.asddls`
- abapGit uses `.asddls`
- vscode syntax highlighting uses `.asddls`, https://github.com/SAP-samples/vscode-abap-cds/blob/main/package.json#L21
- github syntax highligting uses `.asddls`, https://github.com/github/linguist/pull/4614

**Comment by larshp** [comment]
C: its a generic issue, personally I'd like fewer files, everything will show up as changes anyhow

**Comment by BeckerWdf** [comment]
> Currently,
> 
> * ADT uses `.asddls`
> * abapGit uses `.asddls`
> * vscode syntax highlighting uses `.asddls`, https://github.com/SAP-samples/vscode-abap-cds/blob/main/package.json#L21
> * github syntax highligting uses `.asddls`, [Add support for ABAP CDS github/linguist#4614](https://github.com/github/linguist/pull/4614)

Let's look how ABAP classes do it. ADT uses `.aclass` but in this repo classes use `.abap`

File: file-formats/ddls/type/zif_aff_ddls_v1.intf.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://github.com/SAP/abap-file-formats/blob/main/file-formats/intf/intf.json",
```
Comment (huber-nicolas) [inline file-formats/ddls/type/zif_aff_ddls_v1.intf.json:2]: has to refer to the cds schema not the intf schema

**Comment by schneidermic0** [comment]
> abapGit, yes, it uses baseinfo today: https://github.com/abapGit/abapGit/blob/main/src/objects/zcl_abapgit_object_ddls.clas.abap#L236

Then, most probably, the baseinfo is _not_ related to #131.

File: file-formats/ddls/type/zif_aff_ddls_v1.intf.json (hunk 4 above)
Comment (wurzka) [reply file-formats/ddls/type/zif_aff_ddls_v1.intf.json:2]: Not, I think this is correct, as it is the serialized interface for DDLS

**Comment by schneidermic0** [comment]
If possible, I personally would also prefer something like `.cds` or `.ddls` instead of `.asddls`. 

In ADT the file extensions have to be unique per object type to open the correct editor. Therefore, we were not able to use just `.abap`, there.

@BeckerWdf Which file extensions would you use for access controls (DCLS), metadata extension (DDLX), and so on? Would it be the same as for DDLS or a different, because we need other syntax colouring, parsing etc...?

**Comment by BeckerWdf** [comment]
> @BeckerWdf Which file extensions would you use for access controls (DCLS), metadata extension (DDLX), and so on? Would it be the same as for DDLS or a different, because we need other syntax colouring, parsing etc...?

Yes we need another parser but still the language is cds. We would  have:
- <name>.ddls.cds
- <name>.dcls.cds
- <name>.ddlx.cds

So we know if we need to utilise the ddls, dcls or ddlx parser. And there may be other cds parsers that can understand ddls, dcls and ddlx in one parser.

**Review by Christopher-Hermann — APPROVED** [review]
Looks good

**Review by huber-nicolas — APPROVED** [review]

File: file-formats/ddls/ddls.json [hunk 5]
```diff
@@ -0,0 +1,126 @@
+
```
Comment (albertmink) [inline file-formats/ddls/ddls.json:1]: Please remove the empty line.

**Comment by larshp** [comment]
The `.cds` extension does look nicer, but note the impact on the ecosystem

https://cap.cloud.sap/docs/cds/ is the "Language Reference Documentation" for CDS? But is this the same CDS we are looking for?

**Comment by BeckerWdf** [comment]
> If `.cds` files can have different languages, it can be difficult for syntax highlighters to implement. As of right now there are no "ABAP CDS" `.cds` files in the wild

So what would you prefer? `.ddls` or `.ddls.ddls`? First one would be different as the other object types have <object type>.<language>. Second one looks a bit strange because <object type> and <language> are identical.

File: file-formats/ddls/ddls.json (hunk 5 above)
Comment (BeckerWdf) [reply file-formats/ddls/ddls.json]: done

**Comment by larshp** [comment]
I dont really have a good overview

Is ABAP flavored CDS a subset or superset of the other CDS?

But `.acds` could be an option, if its ABAP flavored CDS/ABAP specific CDS

**Comment by BeckerWdf** [comment]
> Is ABAP flavored CDS a subset or superset of the other CDS?

Let's merge this PR and discuss this in https://github.com/SAP/abap-file-formats/issues/161

File: file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json (hunk 3 above)
Comment (BeckerWdf) [reply file-formats/ddls/examples/z_aff_example_ddls.ddls.objectdependencies.json:1]: > Formatting the (.objectdependencies.)json would increase the readability.
Let's merge this PR and work on this in https://github.com/SAP/abap-file-formats/issues/163

**Comment by BeckerWdf** [comment]
Formatting the (.objectdependencies.)json would increase the readability.
Let's merge this PR and work on this in #163

**Pull request merged**